src/main/java/bio/terra/service/filedata/azure/tables/TableDependencyDao.java

```diff
@@ -38,53 +46,59 @@ public void storeSnapshotFileDependencies(
-              // Create any entities that do not already exist
-              List<TableTransactionAction> batchEntities =
-                  refIdChunk.stream()
-                      .distinct()
-                      .filter(id -> !existing.contains(id))
-                      .map(
-                          refId -> {
-                            FireStoreDependency fireStoreDependency =
-                                new FireStoreDependency()
-                                    .snapshotId(snapshotId.toString())
-                                    .fileId(refId)
-                                    .refCount(1L);
-                            TableEntity fireStoreDependencyEntity =
-                                FireStoreDependency.toTableEntity(fireStoreDependency);
-                            return new TableTransactionAction(
-                                TableTransactionActionType.UPSERT_REPLACE,
-                                fireStoreDependencyEntity);
-                          })
-                      .toList();
-              if (!batchEntities.isEmpty()) {
-                // This can happen if retrying a failed transaction.  This would cause an
-                // IllegalArgumentException with message "A transaction must contain at least one
-                // operation"
-                tableClient.submitTransaction(batchEntities);
-              }
-            });
+    List<Future<Void>> futures = new ArrayList<>();
+    for (List<String> refIdChunk : ListUtils.partition(refIds, MAX_FILTER_CLAUSES - 1)) {
+      futures.add(
+          azureTableThreadpool.submit(
+              () -> {
+                String filter =
+                    refIdChunk.stream()
+                            // maybe wrap or cause in parenthesis
+                            .map(
+                                refId ->
+                                    String.format(
+                                        "%s eq '%s'",
+                                        FireStoreDependency.FILE_ID_FIELD_NAME, refId))
+                            .collect(Collectors.joining(" or "))
+                        + String.format(
+                            " and %s eq '%s'",
+                            FireStoreDependency.SNAPSHOT_ID_FIELD_NAME, snapshotId);
+                List<TableEntity> entities =
+                    TableServiceClientUtils.filterTable(
+                        tableServiceClient, dependencyTableName, filter);
+                List<String> existing =
+                    entities.stream()
+                        .map(e -> e.getProperty(FireStoreDependency.FILE_ID_FIELD_NAME).toString())
+                        .toList();
+                // Create any entities that do not already exist
+                List<TableTransactionAction> batchEntities =
+                    refIdChunk.stream()
+                        .distinct()
+                        .filter(id -> !existing.contains(id))
```

Comment:
Given this note, it looks like we're relying on these distinct and filter calls to avoid duplicate entires in the table. It seems like by parallelizing these inserts we could be introducing the chance for duplicates in the table. Should we add some logic before the inserts to check for duplicates in the entries?

Reply:
That's a great point!  I'll distinct the refIds that we get from Synapse

Reply:
Then I can even take out the distinct